File: A4/Python/22101106_CSE221LabAssignment04_Summer2023/Task 08/task8.py
def bfs(graph, start, visited):
    queue = [start]
    visited[start] = 1
    while queue:
        current: int = queue.pop(0)
        for node in graph[current]:
            if visited[node] == -1:
                visited[node] = 1 - visited[current]
                queue.append(node)
            elif visited[node] == visited[current]:
                return False
    return True


with open("input8_1.txt") as f:
    N = int(f.readline())
    cases = ""
    for t in range(N):
        n = int(f.readline())
        graph = [[] for i in range(20001)]
        visited = [-1] * 20001
        for i in range(n):
            u, v = map(int, f.readline().split())
            graph[u].append(v)
            graph[v].append(u)
        ans = 0
        for i in range(1, 20001):
            if visited[i] == -1 and graph[i]:
                if bfs(graph, i, visited):
                    ans += max(visited.count(0), visited.count(1))
                else:
                    ans = -1
                    break
        cases += f"Case {t+1}: {ans}\n"

with open("output8_1.txt", "a+") as out:
    out.truncate(0)
    out.write(cases)

File: A4/Java/22101106_CSE221LabAssignment04_Summer2023/Task_08/task_8.java
package Task_08;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.*;

public class task_8 {

    public static boolean dfs(Map<Integer, List<Integer>> graph, int start, int[] colors) {
        Stack<Integer> stack = new Stack<>();
        stack.push(start);
        colors[start] = 0;

        while (!stack.isEmpty()) {
            int current = stack.pop();
            int currentColor = colors[current];

            for (int neighbor : graph.getOrDefault(current, new ArrayList<>())) {
                if (colors[neighbor] == -1) {
                    colors[neighbor] = 1 - currentColor;
                    stack.push(neighbor);
                } else if (colors[neighbor] == currentColor) {
                    return false;
                }
            }
        }

        return true;
    }

    public static void main(String[] args) {
        for (int i = 1; i < 2; i++) {
            String inputFileName = "Task_08/input8_" + i + ".txt";
            String outputFileName = "Task_08/output8_" + i + ".txt";

            try (BufferedReader br = new BufferedReader(new FileReader(inputFileName));
                 BufferedWriter out = new BufferedWriter(new FileWriter(outputFileName))) {

                int N = Integer.parseInt(br.readLine());
                StringBuilder cases = new StringBuilder();

                for (int t = 0; t < N; t++) {
                    int n = Integer.parseInt(br.readLine());
                    Map<Integer, List<Integer>> graph = new HashMap<>();
                    int[] colors = new int[n + 1];

                    for (int j = 0; j < n; j++) {
                        String[] uv = br.readLine().trim().split(" ");
                        int u = Integer.parseInt(uv[0]);
                        int v = Integer.parseInt(uv[1]);

                        graph.computeIfAbsent(u, k -> new ArrayList<>()).add(v);
                        graph.computeIfAbsent(v, k -> new ArrayList<>()).add(u);
                    }

                    int ans = 0;
                    boolean isValid = true;

                    for (int j = 1; j <= n; j++) {
                        if (colors[j] == -1) {
                            if (dfs(graph, j, colors)) {
                                int group0 = countOccurrences(colors, 0);
                                int group1 = countOccurrences(colors, 1);
                                ans += Math.max(group0, group1);
                            } else {
                                isValid = false;
                                break;
                            }
                        }
                    }

                    if (!isValid) {
                        ans = -1;
                    }

                    cases.append("Case ").append(t + 1).append(": ").append(ans).append("\n");
                }

                out.write(cases.toString());

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static int countOccurrences(int[] arr, int target) {
        int count = 0;
        for (int num : arr) {
            if (num == target) {
                count++;
            }
        }
        return count;
    }
}